### ISSUE TITLE
support for standalone tokenizer

### ISSUE THREAD

**Issue by judaschrist** [elastic/elasticsearch#31880]
**Describe the feature**:
Short version:
Does ES support standalone tokenizers? That is, i want to tokenize my corpus outside ES using my own tool (i.e., python), and then feed the segmented tokens into ES so that i can still use ES's full text search feature, including BM25 based scoring, etc.

Long story:
I am building a text search system where i first use a tokenizer to segment my text corpus, followed by a pipleline of NLP tasks like topic modeling, on independent computing platforms other than ES. I also want to use ES's full text search feature to index my corpus. As a result, the corpus will be tokenized twice, one before the NLP tasks by a python tokenizer, one by my ES tokenizer plugin, which is duplicated and a waste of resource because tokenizing is quite CPU intensive.

Basically I want to be able to only tokenize my corpus just once, and use the result both for NLP tasks and full text search in ES, is this currently supported?

**Comment by danielmitterdorfer** [comment]
At the moment this is not possible. We will be discussing it among the team to get consensus whether we should do this.

**Comment by cbuescher** [comment]
@judaschrist I understand that tokenizing twice feels like a waste of resources of you are already doing it in your pre-processing, but I doubt we will be able to support it since tokenization is an integral part to the Lucene analysis chain. One possible workaround for you might be to do your first tokenization externally (in Python) and then add unique separation markers between each token before sending the text to ES for indexing. You could then make your own ES tokenizer plugin (which already seems to exist) split on exactly those positions (or use e.g. the [Pattern Tokenizer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-pattern-tokenizer.html)) to make use of exactly that tokenization. Scanning the input String  just for detecting split characters shouldn't put much extra demand on resources.

**Comment by cbuescher** [comment]
I relabeled to team-discuss, since its very likely this is mostly interesting to the search & aggs team.

**Comment by judaschrist** [comment]
@cbuescher thanks for the reply. Your workaround solution should do for the moment, will try that out!

**Comment by Tobsucht** [comment]
What about providing such an tokenizer described by @cbuescher in the standard of ES? That would be really useful.

Actually I had the same problem a few months ago. 1) because of the reasons described by @judaschrist and 2) i didn't wanted to write a plugin for such an "easy" use case.

I didn't open an issue at that time because I thought it would contradict with the ideas explained in https://github.com/elastic/elasticsearch/issues/8961 about the updatability of index-time analyzers.

**Comment by cbuescher** [comment]
> the Pattern tokenizer mentioned by @cbuescher is the standard that you're looking for

Exactly.

> i have to insert separation markers into the original text, which will by stored in the _source field which needs some extra coding if you want to display the original text in a searching result.

I also thought more about this after suggesting the workaround. This might actually need some more tricks. I think the easiest would be if you can store the same test field _without_ the tokenization markers in a separate field or as part of a multi-field (doesn't even need to be indexed, you can probably set the field mappings [`enabled`](https://www.elastic.co/guide/en/elasticsearch/reference/current/enabled.html) flag to `false`). The only caveat in this case could be that highlighting might be broken, but it might be worth a try.

**Comment by cbuescher** [comment]
@judaschrist the team discussed the possibility to add accepting external tokanization to Elasticsearch, but we don't see it as something likely to happen in the near future since it has several drawbacks and can be considered something of an expert use-case. If there is increased interest from many sides to implement something like this, we are happy to reconsider. For the time being there is the suggested workaround I mentioned above. Hope this helps in your case.
